Use cycle notation to speed up `permute!`

LilithHafner opened the pull request:
EDIT: nothing in this comment is still true. See the discussion below.

This is an improvement when there are 3+ columns, and should be an improvement for fewer columns once [upstream](https://github.com/scheinerman/Permutations.jl/pull/24) [changes](https://github.com/scheinerman/Permutations.jl/pull/25) merge.

```julia
using DataFrames, Random, BenchmarkTools
df = DataFrame([rand(500_000) for _ in 1:50], :auto);
@benchmark permute!(df, p) setup=(p=shuffle!(collect(1:500_000)))
```
Gives an 8x speedup on my computer: `2.207 s` to `276.050 ms`

(blocked until https://github.com/scheinerman/Permutations.jl/pull/24 merges due to pathological `O(rows^2)` runtime on identity permutation)

bkamins commented:
@scheinerman I have two questions regarding your very nice https://github.com/scheinerman/Permutations.jl package:
* what is the long-term maintenance plan for this package? (the reason I ask is that many packages rely on DataFrames.jl and we try to reduce the number of dependencies of this package as much as possible)
* do you plan adding a function like https://github.com/JuliaData/DataFrames.jl/pull/3035/files#diff-6002b0eaaeb3bc831d7004af3446908da9dd8174c7ec58fedc32565c44935febR2555 in your package? (I ask since it seems to be a natural addition in which case, if we decided to add your package as a dependency it would be probably best to rely on a reference implementation)

@LilithHafner - thank you for the contribution. It looks really cool. We need to resolve the maintenance policy issues listed above. When we are settled how to handle them it would be great to run some benchmarks (especially for cases when permutation creates a lot of short cycles).

CC @nalimilan

nalimilan commented:
Interesting! FWIW, permutation cycles are also implemented in [AbstractAlgebra](https://nemocas.github.io/AbstractAlgebra.jl/latest/perm/), but the package has more dependencies than Permutations.jl.

nalimilan commented on src/abstractdataframe/abstractdataframe.jl:
```diff
@@ -2539,17 +2539,30 @@ function _permutation_helper!(fun::Union{typeof(Base.permute!!), typeof(Base.inv
 
+# permute a vector `v` based on a permutation `p` listed in cycle notation
+function cycle_permute!(v::AbstractVector, cycles::AbstractVector{<:AbstractVector})
+    for cycle in cycles
```
Have you tried `@inbounds` here?

bkamins commented:
Just to make my earlier point clearer:
```
julia> @time cycles(Permutation(1:10^6));
  6.664765 seconds (1.00 M allocations: 93.824 MiB, 1.10% gc time)
```
which is prohibitive I think.

bkamins commented:
Two general comments:
1. for our needs it seems that we can drop cycles of length one (as they are no-op anyway so it is better not to even create vectors for them)
2. we should have a hybrid approach. First we should count number of cycles if it is low (taking into account point 1. above) then we use cycle-based approach; if it is large we use standard implementation.

Given points 1 and 2 what I start to think that maybe the best thing is to port part of the sources to DataFrames.jl and optimize them for our use case without introducing a dependency (of course a proper attribution to the source should be made).

bkamins commented:
maybe even:

3. cycles of length 2 (or other short length) should be also special cased (benchmarking should be done) as tuples with custom swapping code that is an unrolled loop. This will have two benefits: 1. avoid many short loops; 2. avoid allocation of small vectors.

LilithHafner commented:
Permutations.jl currently does not support fast cycle compilation or application. Perhaps down the line, Permutaitons.jl may choose to assimilate these features, but right now it seems like a bad idea to add it as a dependency when it has a lot of unused functionality and none of the functionality we want here.

DataFrame's use case is not particularly special—compile a permutation for fast application and repeatedly apply it—so perhaps it makes sense to factor this out into a package.

@nalimilan I use `@inbounds` in my latest implementation.

Addressing @bkamins's comments:
0) Yes, that is a performance bug fixed by https://github.com/scheinerman/Permutations.jl/pull/24.
1) Yep, my implementation does that
2) I believe a cycle based approach will always be faster, regardless of the number of cycles, the reason Permutations.jl is very slow on the identity permutation is because it runs in O(n^2) when O(n) is possible due to repeated `findfirst`. (see 0).
3) This seems like more work than it's worth, and my current implementation only allocates once anyway, inlining all the cycles.

LilithHafner commented:
For benchmarks I ran the following:
```julia
using DataFrames, Random, BenchmarkTools
for order in [identity, reverse, shuffle]
    for rows in [10, 100, 100_000]
        for columns in [1,2,3,5,50]
            df = DataFrame([rand(rows) for _ in 1:columns], :auto);
            t = @belapsed permute!($df, p) setup=(p=$order(collect(1:$rows)))
            println("order = $order, rows = $rows, columns = $columns: time = $t")
        end
    end
end
```
<details>
  <summary>Formatted results:</summary>
  
 ```
order = identity, rows = 10,     columns = 1:  time = 3.4898104265402846e-7 -> 3.920398009950249e-7 (ratio = 1.1233842331764818)
order = identity, rows = 10,     columns = 2:  time = 4.5393401015228427e-7 -> 4.614183673469388e-7 (ratio = 1.0164877648011958)
order = identity, rows = 10,     columns = 3:  time = 5.886646706586827e-7 -> 5.232947368421053e-7  (ratio = 0.8889521707775803)
order = identity, rows = 10,     columns = 5:  time = 7.870520833333334e-7 -> 6.5046875e-7          (ratio = 0.8264621411649482)
order = identity, rows = 10,     columns = 50: time = 8.753e-6 -> 7.566666666666667e-6              (ratio = 0.86446551658479)
order = identity, rows = 100,    columns = 1:  time = 6.771184210526315e-7 -> 6.045588235294117e-7  (ratio = 0.8928406091649073)
order = identity, rows = 100,    columns = 2:  time = 1.1778e-6 -> 6.856778523489932e-7             (ratio = 0.5821683242901964)
order = identity, rows = 100,    columns = 3:  time = 1.572e-6 -> 7.524210526315789e-7              (ratio = 0.4786393464577474)
order = identity, rows = 100,    columns = 5:  time = 2.4561111111111115e-6 -> 9.18469387755102e-7  (ratio = 0.3739527025467504)
order = identity, rows = 100,    columns = 50: time = 2.3675e-5 -> 6.1304e-6                        (ratio = 0.25893980992608234)
order = identity, rows = 100000, columns = 1:  time = 0.00045358 -> 0.000212106                     (ratio = 0.46762643855549185)
order = identity, rows = 100000, columns = 2:  time = 0.000981642 -> 0.000211538                    (ratio = 0.21549403957858362)
order = identity, rows = 100000, columns = 3:  time = 0.001469507 -> 0.000217253                    (ratio = 0.14784073842451922)
order = identity, rows = 100000, columns = 5:  time = 0.00250761 -> 0.000216133                     (ratio = 0.0861908350979618)
order = identity, rows = 100000, columns = 50: time = 0.025942932 -> 0.000258782                    (ratio = 0.00997504830988263)
order = reverse,  rows = 10,     columns = 1:  time = 4.0316582914572867e-7 -> 3.9386597938144333e-7(ratio = 0.9769329415045147)
order = reverse,  rows = 10,     columns = 2:  time = 4.849128205128206e-7 -> 4.6842564102564105e-7 (ratio = 0.9659997038854459)
order = reverse,  rows = 10,     columns = 3:  time = 6.30022346368715e-7 -> 5.546075268817205e-7   (ratio = 0.8802981831967296)
order = reverse,  rows = 10,     columns = 5:  time = 7.877631578947369e-7 -> 7.307460317460317e-7  (ratio = 0.9276214867663005)
order = reverse,  rows = 10,     columns = 50: time = 7.26625e-6 -> 7.637e-6                        (ratio = 1.0510235678651298)
order = reverse,  rows = 100,    columns = 1:  time = 5.89585798816568e-7 -> 7.560092592592593e-7   (ratio = 1.282271826724356)
order = reverse,  rows = 100,    columns = 2:  time = 9.832e-7 -> 8.94030303030303e-7               (ratio = 0.9093066548314718)
order = reverse,  rows = 100,    columns = 3:  time = 1.2844000000000001e-6 -> 1.2245999999999999e-6(ratio = 0.9534412955465585)
order = reverse,  rows = 100,    columns = 5:  time = 6.753555555555556e-6 -> 1.3894e-6             (ratio = 0.20572866967194234)
order = reverse,  rows = 100,    columns = 50: time = 6.761e-5 -> 1.3521e-5                         (ratio = 0.19998520928856678)
order = reverse,  rows = 100000, columns = 1:  time = 0.001034277 -> 0.000436326                    (ratio = 0.42186570908953797)
order = reverse,  rows = 100000, columns = 2:  time = 0.00207731 -> 0.000562078                     (ratio = 0.27057974014470637)
order = reverse,  rows = 100000, columns = 3:  time = 0.00306503 -> 0.000732566                     (ratio = 0.23900777480155172)
order = reverse,  rows = 100000, columns = 5:  time = 0.005106755 -> 0.001039143                    (ratio = 0.20348401284181442)
order = reverse,  rows = 100000, columns = 50: time = 0.058325733 -> 0.00757876                     (ratio = 0.1299385298766841)
order = shuffle, rows = 10,     columns = 1:  time = 1.198e-6 -> 4.0047738693467337e-7              (ratio = 0.3342883029504786)
order = shuffle, rows = 10,     columns = 2:  time = 1.5404000000000001e-6 -> 4.97876923076923e-7   (ratio = 0.32321275193256493)
order = shuffle, rows = 10,     columns = 3:  time = 1.8994000000000002e-6 -> 5.52731182795699e-7   (ratio = 0.29100304453811676)
order = shuffle, rows = 10,     columns = 5:  time = 2.6708888888888887e-6 -> 7.468679245283019e-7  (ratio = 0.27963271989161814)
order = shuffle, rows = 10,     columns = 50: time = 3.1753e-5 -> 6.2836e-6                         (ratio = 0.1978899631530879)
order = shuffle, rows = 100,    columns = 1:  time = 1.9177e-6 -> 7.042361111111111e-7              (ratio = 0.3672295516040628)
order = shuffle, rows = 100,    columns = 2:  time = 3.117125e-6 -> 8.928571428571428e-7            (ratio = 0.286436104698125)
order = shuffle, rows = 100,    columns = 3:  time = 4.09e-6 -> 1.0733e-6                           (ratio = 0.2624205378973105)
order = shuffle, rows = 100,    columns = 5:  time = 6.151e-6 -> 1.4070999999999998e-6              (ratio = 0.22875955129247275)
order = shuffle, rows = 100,    columns = 50: time = 5.7618e-5 -> 1.0821e-5                         (ratio = 0.187805893991461)
order = shuffle, rows = 100000, columns = 1:  time = 0.001764317 -> 0.001812535                     (ratio = 1.027329555856459)
order = shuffle, rows = 100000, columns = 2:  time = 0.004136857 -> 0.002170336                     (ratio = 0.5246340398036481)
order = shuffle, rows = 100000, columns = 3:  time = 0.005132534 -> 0.002502598                     (ratio = 0.4875950164187904)
order = shuffle, rows = 100000, columns = 5:  time = 0.009843017 -> 0.003137679                     (ratio = 0.31877207973937255)
order = shuffle, rows = 100000, columns = 50: time = 0.126754584 -> 0.021798389                     (ratio = 0.1719731808673681)
```
</details>

bkamins commented on src/abstractdataframe/abstractdataframe.jl:
```diff
@@ -2539,17 +2539,72 @@ function _permutation_helper!(fun::Union{typeof(Base.permute!!), typeof(Base.inv
+    count = length(p)
+    @inbounds while count > 0
+        start = findnext(!iszero, p, start+1)
+        start isa Int || throw(ArgumentError("Not a permutation"))
```
would `start === nothing &&` be faster or it is the same?

LilithHafner replied:
Same

bkamins commented on src/abstractdataframe/abstractdataframe.jl:
```diff
@@ -2539,17 +2539,72 @@ function _permutation_helper!(fun::Union{typeof(Base.permute!!), typeof(Base.inv
+# destroying the original permutation in the process.
+function _compile_permutation!(p::AbstractVector{<:Integer})
+    Base.require_one_based_indexing(p)
+    out = similar(p, 3length(p)÷2)
```
```suggestion
    out = similar(p, 3 * length(p) ÷ 2)
```

bkamins replied:
also maybe add a comment that this length is guaranteed to be enough because we are not recording length 1 cycles.

bkamins commented on src/abstractdataframe/abstractdataframe.jl:
```diff
@@ -2539,17 +2539,72 @@ function _permutation_helper!(fun::Union{typeof(Base.permute!!), typeof(Base.inv
+    start = 0
+    count = length(p)
+    @inbounds while count > 0
+        start = findnext(!iszero, p, start+1)
```
```suggestion
        start = findnext(!iszero, p, start + 1)
```

bkamins commented on src/abstractdataframe/abstractdataframe.jl:
```diff
@@ -2539,17 +2539,72 @@ function _permutation_helper!(fun::Union{typeof(Base.permute!!), typeof(Base.inv
+        k = p[start]
+        count -= 1
+        k == start && continue
+        out[out_len += 1] = k
```
can you pleaes split here and below `out[out_len += 1] = k` into two lines?

LilithHafner replied:
done

bkamins commented on src/abstractdataframe/abstractdataframe.jl:
```diff
@@ -2539,17 +2539,72 @@ function _permutation_helper!(fun::Union{typeof(Base.permute!!), typeof(Base.inv
+        v[last_p_i] = start
+        i += 1
+    end
+    v
```
```suggestion
    return v
```

bkamins commented on src/abstractdataframe/abstractdataframe.jl:
```diff
@@ -2539,17 +2539,72 @@ function _permutation_helper!(fun::Union{typeof(Base.permute!!), typeof(Base.inv
+            k == 0 && break # or k < start+1
+        end
+    end
+    resize!(out, out_len)
```
```suggestion
    return resize!(out, out_len)
```

bkamins commented:
Thank you! I have left some stylistic convention comments that we follow in DataFrames.jl. I will test it more later.

As a comment for the future. In sorting code we create a standard permutation and then pass it to `permute!`. Maybe we could make a change there not to create an intermediate vector and create the cycles on the first pass? (this is for a separate PR and I am not sure how easy this would be)

bkamins commented on src/abstractdataframe/abstractdataframe.jl:
```diff
@@ -2539,17 +2539,72 @@ function _permutation_helper!(fun::Union{typeof(Base.permute!!), typeof(Base.inv
 
+# convert a classical permutation to cycle notation, 
+# destroying the original permutation in the process.
+function _compile_permutation!(p::AbstractVector{<:Integer})
```
is there any cease when we do not pass `AbstractVector{Int}` here? Just wondering, as e.g. `<:Integer` allows `Bool`.

bkamins commented on src/abstractdataframe/abstractdataframe.jl:
```diff
@@ -2539,17 +2539,72 @@ function _permutation_helper!(fun::Union{typeof(Base.permute!!), typeof(Base.inv
+# convert a classical permutation to cycle notation, 
+# destroying the original permutation in the process.
+function _compile_permutation!(p::AbstractVector{<:Integer})
+    Base.require_one_based_indexing(p)
```
can you please change it to test using `firstindex` as this fails on Julia 1.0?

LilithHafner replied:
Done

LilithHafner commented on src/abstractdataframe/abstractdataframe.jl:
```diff
@@ -2539,17 +2539,72 @@ function _permutation_helper!(fun::Union{typeof(Base.permute!!), typeof(Base.inv
+            last_k = k
+            checkbounds(Bool, p, k) || throw(ArgumentError("Not a permutation"))
+            k = out[out_len += 1] = p[k]
+            k == 0 && break # or k < start+1
```
It means that that condition (`k == 0`) could also be `k < start+1`. We could tell this to the compiler with `@assert `(k == 0) == (k < start + 1) == (k < start)`, but I don't think that's worth the confusion.

I'll benchmark equivalent tests, when they inevitably come up identical I'll delete the comment.

LilithHafner replied:
Update, it's all the same performance-wise and `k == 0` is the clearest. I've removed the comment.

bkamins commented on src/abstractdataframe/abstractdataframe.jl:
```diff
@@ -2533,23 +2533,80 @@ function _permutation_helper!(fun::Union{typeof(Base.permute!!), typeof(Base.inv
+    return resize!(out, out_len)
+end
+
+# permute a vector `v` based on a permutation `p` listed in cycle notation
```
Can you please add a comment about the expected format of cycle notation? (i.e. cycles followed by 0 all encoded in a single vector)?

bkamins commented on src/abstractdataframe/abstractdataframe.jl:
```diff
@@ -2588,7 +2588,10 @@ function _compile_permutation!(p::AbstractVector{<:Integer})
+# Permute a vector `v` based on a permutation `p` listed in zero terminated
+# cycle notation. For example, the permutation 1 -> 2, 2 -> 3, 3 -> 1, 4 -> 6,
+# 5 -> 5, 6 -> 4 is traditionaly expressed as [2, 3, 1, 6, 5, 4] but in cycle
+# notation is expressed as [1, 2, 3, 0, 4, 6, 0] (or [1, 2, 3, 0, 4, 6, 0, 5, 0])
```
```suggestion
# notation used in the implementation is expressed as [1, 2, 3, 0, 4, 6, 0]
```
I would remove the alternative as we do not use it

LilithHafner replied:
okay

bkamins commented on NEWS.md:
```diff
@@ -26,6 +26,8 @@
 * Add `fillcombinations` function that generates all combinations of
   levels of selected columns of a data frame
   ([#3012](https://github.com/JuliaData/DataFrames.jl/issues/3012))
+* Guarantee that permute! and invpermute! throw on invalid input
```
```suggestion
* Guarantee that `permute!` and `invpermute!` throw on invalid input
```

bkamins commented on NEWS.md:
```diff
@@ -34,6 +36,12 @@
 
+## Performance
+
+* Speed up permute! and invpermute! (and therefore sort!) 2x-8x 
```
```suggestion
* Speed up `permute!` and `invpermute!` (and therefore sort!) 2x-8x 
```

bkamins commented:
>  Do you also want to include brute-force testing for this in CI?

Yes please. They should take no more than several seconds and experience shows that they allow catching things in the future (i.e. even if things work now someone in the future might want to change something in the code and accidentally break things).

LilithHafner commented:
Being a bit more thorough with exhaustive tests did catch something! `invperm!(df, [])` used to fail, now it works.

bkamins commented on src/abstractdataframe/abstractdataframe.jl:
```diff
@@ -2531,33 +2531,95 @@ function _permutation_helper!(fun::Union{typeof(Base.permute!!), typeof(Base.inv
         if haskey(seen_cols, col)
             push!(toskip, i)
         else
+            checkbounds(col, eachindex(p))
```
why do you think it is needed? This part of code can assume that each column in a data frame is 1-based and has equal length. We check that this invariant is met elsewhere. We might just check if length of `p` equals number of rows in `df`.

LilithHafner replied:
Thanks! I did not know that we can assume the dataframe is one-indexed and every column has the same length. I removed that check and replaced it with a check that `length(p) <= nrow(df)` because short permutations are supported.

EDIT: moved from main thread

LilithHafner commented:
Thanks! I did not know that we can assume the dataframe is one-indexed and every column has the same length. I removed that check and replaced it with a check that `length(p) <= nrow(df)` because short permutations are supported.

bkamins commented on src/abstractdataframe/abstractdataframe.jl:
```diff
@@ -2525,39 +2525,95 @@ end
 
-To return a new data frame instead of permuting `df` in-place, use `df[p]`.
-Note that this is generally faster than `permute!(df, p)` for large data frames.
+To return a new data frame instead of permuting `df` in-place, use `df[p, :]`.
```
While we are at it can you please also correct line 2621 in this PR (`permute!` should be `invpermute!` there). Thank you

LilithHafner replied:
Done

bkamins commented on src/abstractdataframe/abstractdataframe.jl:
```diff
@@ -2525,39 +2525,95 @@ end
 function _permutation_helper!(fun::Union{typeof(Base.permute!!), typeof(Base.invpermute!!)},
                               df::AbstractDataFrame, p::AbstractVector{<:Integer})
-    toskip = Set{Int}()
+    length(p) <= nrow(df) || throw(ArgumentError("Permutation is too long"))
```
@nalimilan - what do you think we have two options:
1. leave what @LilithHafner proposes - then we need to update the docstring explaining that shorter permutations are allowed.
2. do `==` test and require permutation to have an exact length?

LilithHafner replied:
Oh, I misread the tests and thought that we already supported small permutations previously and needed to maintain that. I don't see much benefit to supporting short vectors, especially if they have one-based indexing. Seems better to throw an error. Aside: In the vein of short vectors, it might be nice to specialize on sparse vectors, but this PR already does a good job of handling almost-identity transformations efficiently.

I vote 2.

bkamins replied:
OK - so can you please change it to 2 (i.e. require `==`)

bkamins replied:
I have pushed the change to approve the PR.

bkamins commented on src/abstractdataframe/abstractdataframe.jl:
```diff
@@ -2525,39 +2525,95 @@ end
 function _permutation_helper!(fun::Union{typeof(Base.permute!!), typeof(Base.invpermute!!)},
                               df::AbstractDataFrame, p::AbstractVector{<:Integer})
-    toskip = Set{Int}()
+    length(p) <= nrow(df) || throw(ArgumentError("Permutation is too long"))
```
```suggestion
    length(p) == nrow(df) || throw(ArgumentError("Permutation is too long"))
```

bkamins commented on src/abstractdataframe/abstractdataframe.jl:
```diff
@@ -2525,39 +2525,95 @@ end
 function _permutation_helper!(fun::Union{typeof(Base.permute!!), typeof(Base.invpermute!!)},
                               df::AbstractDataFrame, p::AbstractVector{<:Integer})
-    toskip = Set{Int}()
+    length(p) == nrow(df) || throw(ArgumentError("Permutation is too long"))
```
```suggestion
    length(p) == nrow(df) || throw(ArgumentError("Permutation does not have a correct length"))
```

bkamins commented on src/abstractdataframe/abstractdataframe.jl:
```diff
@@ -2525,39 +2525,95 @@ end
+    isempty(cp) && return df
+
+    if fun === Base.invpermute!! 
+        pop!(cp)
+        reverse!(cp)
+        push!(cp, 0)
```
Interestingly it is roughly equally fast as `reverse!(@view cp[1:end-1])`

LilithHafner replied:
Great! I've verified that with a microbenchmark that finds `reverse!` slightly faster. I'm guessing the compiler makes the view transparent. I pushed that change.

bkamins replied:
Thank you (I have not pushed for this as the difference was minimal 😄)

bkamins commented on test/dataframe.jl:
```diff
@@ -2165,45 +2165,47 @@ end
+                    @test_throws DimensionMismatch("Permutation does not have a correct length ($len != $perm_len)") invpermute!(df, p)
                 elseif sort(p) != collect(1:perm_len)
-                    if len < 5
+                    if perm_len <= 4 || rand() < .005
```
With random tests we typically set a seed explicitly in the top of the `@testset` to make sure all is easily reproducible.

LilithHafner replied:
Done

bkamins commented on src/abstractdataframe/abstractdataframe.jl:
```diff
@@ -2552,8 +2552,10 @@ function _compile_permutation!(p::AbstractVector{<:Integer})
-    out = similar(p, 3 * length(p) ÷ 2) 
+    # so the worst case is all 2-cycles. One extra element gives the
+    # algorithm leeway to defer error detection without unsafe reads.
+    # trace _compile_permutation!([3,3,1]) for example.
```
This was the reason of my original comment. So in fact you used this one extra element for 1-element cycle temporarily although you do not store it😄 (initially I thought you do not use it).

LilithHafner replied:
Your initial thought is correct for inputs that are permutations. Legitimate one-cycles are filtered immediately by `last_k == start && continue` and not even stored temporarily in the output vector.

The issue is that a faulty input vector can bypass this check one time, and we need to throw an ArgumentError before performing an unsafe write. For example, on input [3, 3, 1] (not a permutation), we first identify and store a 2-cycle as [3, 1, 0], and then we see the next three as the start of a non-one-cycle (because it is a 3 at the 2nd index) and store it as [3, 1, 0, 3].  We then write the next element in the cycle (0, because the value at spot 3 in the input has been zeroed out by the first cycle), so we get [3, 1, 0, 3, 0], and only then do we recognize that something is wrong because we ended a cycle somewhere other than where it started and we throw an error with `last_k == start  || throw(ArgumentError("Not a permutation"))`. But at this point, unless we have 5 spots preallocated in `out`, we have already over-written.

bkamins replied:
Yes. Just to be sure. For invalid permutations it can happen only at most once - right?

LilithHafner replied:
I hope so :) Every element that goes from the input vector into a cycle gets zeroed out in the input vector, so a 1-cycle can masquerade as the beginning of a nontrivial cycle, but will have to either point out of bounds (checked immediately), to an as yet untransferred entry in the input (in this case there will be room left in the output vector), or to an already transferred element in which case the loop will terminate and throw after one extra write.

bkamins commented on test/dataframe.jl:
```diff
@@ -2187,7 +2188,7 @@ end
                     @test_throws DimensionMismatch("Permutation does not have a correct length ($len != $perm_len)") invpermute!(df, p)
                 elseif sort(p) != collect(1:perm_len)
-                    if perm_len <= 4 || rand() < .005
+                    if perm_len <= 5 || rand() < .01
```
maybe just check it always, i.e. remove the `rand() < 0.01` (what is the benefit of keeping this additional condition?)

LilithHafner replied:
The benefit is runtime. Without the check runtime jumps from 6 seconds to 143 seconds.

LilithHafner replied:
40,661

bkamins replied:
OK - this is enough :)

bkamins commented on test/dataframe.jl:
```diff
@@ -2163,6 +2164,50 @@ end
+                    @test_throws DimensionMismatch("Permutation does not have a correct length ($len != $perm_len)") permute!(df, p)
+                    @test_throws DimensionMismatch("Permutation does not have a correct length ($len != $perm_len)") invpermute!(df, p)
+                elseif sort(p) != collect(1:perm_len)
+                    if perm_len <= 5 || rand() < .01
```
```suggestion
                    if perm_len <= 5 || rand() < 0.01
```
some editors get confused by `.01` when highlighting code.

bkamins commented:
@nalimilan - could you please have a look at this PR before I merge it? Thank you!

LilithHafner commented:
I'm assuming I don't need to worry about the out-of-memory error on thread tests on windows

nalimilan commented on src/abstractdataframe/abstractdataframe.jl:
```diff
@@ -2525,39 +2525,95 @@ end
 function _permutation_helper!(fun::Union{typeof(Base.permute!!), typeof(Base.invpermute!!)},
                               df::AbstractDataFrame, p::AbstractVector{<:Integer})
-    toskip = Set{Int}()
+    nrow(df) != length(p) && throw(DimensionMismatch("Permutation does not have a correct length ($(nrow(df)) != $(length(p)))"))
```
```suggestion
    nrow(df) != length(p) &&
        throw(DimensionMismatch("Permutation does not have a correct length " *
                                "(expected $(nrow(df)) but got $(length(p)))"))
```

nalimilan commented on src/abstractdataframe/abstractdataframe.jl:
```diff
@@ -2525,39 +2525,95 @@ end
+    count = length(p)
+    @inbounds while count > 0
+        start = findnext(!iszero, p, start + 1)
+        start isa Int || throw(ArgumentError("Not a permutation"))
```
This may not be super explicit for somebody calling `permute!`. Maybe say something like "Passed vector `p` is not a valid permutation"?

bkamins replied:
I have incorporated all your suggestions in the PR. Once the CI passes I will merge it.

nalimilan commented on src/abstractdataframe/abstractdataframe.jl:
```diff
@@ -2525,39 +2525,95 @@ end
-            fun(col, pp)
+# Permute a vector `v` based on a permutation `p` listed in zero terminated
+# cycle notation. For example, the permutation 1 -> 2, 2 -> 3, 3 -> 1, 4 -> 6,
+# 5 -> 5, 6 -> 4 is traditionaly expressed as [2, 3, 1, 6, 5, 4] but in cycle
```
```suggestion
# 5 -> 5, 6 -> 4 is traditionally expressed as [2, 3, 1, 6, 5, 4] but in cycle
```

nalimilan reviewed (commented):
I haven't followed the technical discussions, but if you think it's correct that's fine with me!

bkamins commented:
> I'm assuming I don't need to worry about the out-of-memory error on thread tests on windows

No - this is a problem on 32 bit machine that sometimes does not get enough RAM in CI which we cannot control.

bkamins commented on src/abstractdataframe/abstractdataframe.jl:
```diff
@@ -2525,39 +2525,97 @@ end
+    count = length(p)
+    @inbounds while count > 0
+        start = findnext(!iszero, p, start + 1)
+        start isa Int || throw(ArgumentError("Not a permutation"))
```
```suggestion
        start isa Int || throw(ArgumentError("Passed vector p is not a valid permutation"))
```

bkamins commented on src/abstractdataframe/abstractdataframe.jl:
```diff
@@ -2525,39 +2525,97 @@ end
+        out_len += 1
+        out[out_len] = last_k
+        p[start] = 0
+        start <= last_k <= length(p) || throw(ArgumentError("Not a permutation"))
```
```suggestion
        start <= last_k <= length(p) || throw(ArgumentError("Passed vector p is not a valid permutation"))
```

bkamins commented on src/abstractdataframe/abstractdataframe.jl:
```diff
@@ -2525,39 +2525,97 @@ end
+            count -= 1
+            p[last_k] = 0
+            last_k = k
+            start <= k <= length(p) || throw(ArgumentError("Not a permutation"))
```
```suggestion
            start <= k <= length(p) || throw(ArgumentError("Passed vector p is not a valid permutation"))
```

bkamins commented on src/abstractdataframe/abstractdataframe.jl:
```diff
@@ -2525,39 +2525,97 @@ end
+            k = out[out_len] = p[k]
+            k == 0 && break
+        end
+        last_k == start  || throw(ArgumentError("Not a permutation"))
```
```suggestion
        last_k == start  || throw(ArgumentError("Passed vector p is not a valid permutation"))
```

bkamins commented on src/abstractdataframe/abstractdataframe.jl:
```diff
@@ -2525,39 +2525,97 @@ end
+        last_k == start && continue
+        out_len += 1
+        out[out_len] = last_k
+        p[start] = 0
+        start <= last_k <= length(p) || throw(ArgumentError("Passed vector p is not a valid permutation"))
```
```suggestion
        start < last_k <= length(p) || throw(ArgumentError("Passed vector p is not a valid permutation"))
```

bkamins replied:
@LilithHafner - I have added a suggestion that my code analysis shows that are valid. Can you please confirm before I merge the PR that:
1. we do not need to set `p[start]` to `0` as we never go back to check the value of `p[start]` in the algorithm (actually you take advantage of this fact already in line `last_k == start && continue` if I understand your code correctly).
2. `last_k` cannot be equal to `start` here.

This change does not change the result of the code, but I just wanted to ensure that I understand your code correctly. The only benefit is that the condition would be a bit tighter + we would have some small performance gain (most visible if permutation consists only of 2-cycles).

bkamins replied:
@LilithHafner so the `p[start] = 0` part is needed here because of `k = out[out_len] = p[k]`. Right? This makes sense.

LilithHafner replied:
1. No. Without any other changes, removing `p[start] = 0` will cause problems because cycle termination detection is done by checking that `k == 0`, so if the starting element is not zeroed out before, it will be zeroed out as the last element of the cycle, thus double counting it.

2. Yes. If `last_k == start`, we've got a problem slightly more serious than the passed vector is not a permutation, but a stricter condition is nice.

bkamins replied:
> will cause problems because cycle termination detection is done by checking that `k == 0`

This is what I meant. Thank you for confirming.

LilithHafner replied:
Oops! I was out of sync and posted before I saw your post :) I'm glad we both drew the same conclusion.

bkamins commented on src/abstractdataframe/abstractdataframe.jl:
```diff
@@ -2525,39 +2525,97 @@ end
+            count -= 1
+            p[last_k] = 0
+            last_k = k
+            start <= k <= length(p) || throw(ArgumentError("Passed vector p is not a valid permutation"))
```
```suggestion
            start < k <= length(p) || throw(ArgumentError("Passed vector p is not a valid permutation"))
```

bkamins commented on src/abstractdataframe/abstractdataframe.jl:
```diff
@@ -2525,39 +2525,97 @@ end
+            k = out[out_len] = p[k]
+            k == 0 && break
+        end
+        last_k == start  || throw(ArgumentError("Passed vector p is not a valid permutation"))
```
```suggestion
```

bkamins commented on test/dataframe.jl:
```diff
@@ -2163,6 +2164,50 @@ end
+
+    @test DataFrame(x=[17]) == invpermute!(DataFrame(x=[17]), [1])
+
+    @test_throws DimensionMismatch("Permutation does not have a correct length (5 != 6)") permute!(df, [1, 4, 3, 2, 5, 6])
+    @test_throws DimensionMismatch("Permutation does not have a correct length (5 != 3)") permute!(df, [1, 3, 2])
+    @test_throws ArgumentError("Not a permutation") permute!(df, [4, 4, 2, 5, 3])
+    @test_throws ArgumentError("Not a permutation") DataFrames._compile_permutation!([3, 3, 1])
+    @test_throws ArgumentError("Permutation vectors must have 1-based indexing") invpermute!(df, OffsetArray([5,4,7,8,6], 4:8))
```
```suggestion
    @test_throws DimensionMismatch permute!(df, [1, 4, 3, 2, 5, 6])
    @test_throws DimensionMismatch permute!(df, [1, 3, 2])
    @test_throws ArgumentError permute!(df, [4, 4, 2, 5, 3])
    @test_throws ArgumentError DataFrames._compile_permutation!([3, 3, 1])
    @test_throws ArgumentError invpermute!(df, OffsetArray([5,4,7,8,6], 4:8))
```

bkamins commented on test/dataframe.jl:
```diff
@@ -2163,6 +2164,50 @@ end
+                digits!(p, i, base=perm_len+3)
+                p .-= 2
+                if perm_len != len
+                    @test_throws DimensionMismatch("Permutation does not have a correct length ($len != $perm_len)") permute!(df, p)
+                    @test_throws DimensionMismatch("Permutation does not have a correct length ($len != $perm_len)") invpermute!(df, p)
```
```suggestion
                    @test_throws DimensionMismatch permute!(df, p)
                    @test_throws DimensionMismatch invpermute!(df, p)
```

LilithHafner commented:
Looks good to me!

LilithHafner commented on test/dataframe.jl:
```diff
@@ -2163,6 +2164,51 @@ end
+        len = rand(1:1000)
+        p = shuffle!(collect(1:len))
+        v = rand(len)
+        df = DataFrame([v, v, rand(len), v], :auto)
+        dfc = copy(df)
+        @test df[p, :] == permute!(df, p)
+        @test dfc == invpermute!(df, p)
```
Thanks!

bkamins replied:
(the point was to check if we correctly handle aliasing)

bkamins commented:
👏 @LilithHafner 

Thank you!

LilithHafner commented:
@bkamins, thank you for your review!

scheinerman commented:
> @scheinerman I have two questions regarding your very nice https://github.com/scheinerman/Permutations.jl package:
> 
> * what is the long-term maintenance plan for this package? (the reason I ask is that many packages rely on DataFrames.jl and we try to reduce the number of dependencies of this package as much as possible)
> * do you plan adding a function like https://github.com/JuliaData/DataFrames.jl/pull/3035/files#diff-6002b0eaaeb3bc831d7004af3446908da9dd8174c7ec58fedc32565c44935febR2555 in your package? (I ask since it seems to be a natural addition in which case, if we decided to add your package as a dependency it would be probably best to rely on a reference implementation)
> 
> @LilithHafner - thank you for the contribution. It looks really cool. We need to resolve the maintenance policy issues listed above. When we are settled how to handle them it would be great to run some benchmarks (especially for cases when permutation creates a lot of short cycles).
> 
> CC @nalimilan

Hello. I don't have a long-term maintenance plan. Suggestions?

@LilithHafner You suggested adding a function and gave me some sample code. Perhaps it would be easiest if you did that and submitted a pull request? Or is there a better way to handle that?

LilithHafner commented:
If you're interested in hosting compiled permutations in permutations.jl (which seems like a reasonable choice) then I'll make a PR there. At some point DataFrames.jl might start using Permutaitons.jl, or not. It's also plausible that there should be slight differences between the two packages' implementations because of differences in package scope and focus.

bkamins commented:
@scheinerman - I 100% agree with @LilithHafner. For now we have our custom implementation merged to DataFrames.jl. I think it would be great to have a general implementation in Permutations.jl. Then at some point we might check if it is worth to leave the things as they are right now or integrate the packages. Thank you!

scheinerman commented:
I see that @LilithHafner added items to the code in `Permutations` via two pull requests, and I was happy to merge them in.
